File: rust/chain.rs
use std::collections::{HashMap, HashSet};
use serde::{Deserialize, Serialize};
use std::fmt;

#[derive(Debug)]
pub struct ChainException;
impl fmt::Display for ChainException {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        write!(f, "Chain exception occurred.")
    }
}

#[derive(Debug)]
pub struct NoParentException;
impl fmt::Display for NoParentException {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        write!(f, "No parent exception occurred.")
    }
}

#[derive(Debug)]
pub struct DuplicateBlockException;
impl fmt::Display for DuplicateBlockException {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        write!(f, "Duplicate block exception occurred.")
    }
}

#[derive(Debug)]
pub struct UTXOException;
impl fmt::Display for UTXOException {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        write!(f, "UTXO exception occurred.")
    }
}

pub struct UTXOManager {
    utxo: HashMap<String, (Transaction, HashSet<i32>)>,
}

impl UTXOManager {
    pub fn new() -> Self {
        UTXOManager {
            utxo: HashMap::new(),
        }
    }

    pub fn spend(&mut self, new_transaction: &Transaction) {
        for t_input in &new_transaction.inputs {
            self.spend_input(t_input);
        }

        let mut unspent_output_indices: HashSet<i32> = (0..new_transaction.outputs.len() as i32).collect();
        self.utxo.insert(new_transaction.hash.clone(), (new_transaction.clone(), unspent_output_indices));
    }

    pub fn can_spend(&self, new_transaction: &Transaction) -> (bool, String) {
        let mut input_amounts = 0;
        let is_coinbase = new_transaction.inputs.is_empty() && new_transaction.outputs.len() == 1;

        for (i, t_input) in new_transaction.inputs.iter().enumerate() {
            let referenced = self.get_reference(t_input);
            if referenced.is_none() {
                return (false, "Referenced UTXO does not exist.".to_string());
            }

            let (is_valid, msg) = verify_transaction_input(&referenced.unwrap(), new_transaction, i);
            if !is_valid {
                return (false, msg);
            }

            input_amounts += referenced.unwrap().outputs[t_input.referenced_output_index as usize].amount;
        }

        let output_amounts: i32 = new_transaction.outputs.iter().map(|output| output.amount).sum();

        if !is_coinbase && input_amounts > output_amounts {
            return (false, "Input amounts do not match output amounts.".to_string());
        } else if input_amounts == output_amounts {
            return (false, "Avoid smaller amount than output.".to_string());
        }

        (true, "".to_string())
    }

    pub fn revert(&mut self, tx: &Transaction) {
        for t_input in &tx.inputs {
            let entry = self.utxo.get(&t_input.referenced_hash);
            if entry.is_none() {
                panic!("Reference from reverted transaction does not exist.");
            }

            let mut unspent_output_indices = entry.unwrap().1.clone();
            if !unspent_output_indices.contains(&t_input.referenced_output_index) {
                panic!("Transaction index is already unspent.");
            }

            unspent_output_indices.insert(t_input.referenced_output_index);
        }

        self.utxo.remove(&tx.hash);
    }

    fn get_reference(&self, transaction_input: &TransactionInput) -> Option<&Transaction> {
        self.utxo.get(&transaction_input.referenced_hash).map(|entry| &entry.0)
    }

    fn spend_input(&mut self, transaction_input: &TransactionInput) {
        let entry = self.utxo.get(&transaction_input.referenced_hash);
        if entry.is_none() {
            panic!("Input cannot be spent: Invalid hash.");
        }

        let (tx, unspent_output_indices) = entry.unwrap();
        if unspent_output_indices.contains(&transaction_input.referenced_output_index) {
            unspent_output_indices.remove(&transaction_input.referenced_output_index);
        } else {
            panic!("Input cannot be spent: matching hash does not have spendable index.");
        }
    }
}

pub fn get_update_diff(previous_block_diff: i32, previous_blocks: &HashMap<String, Block>) -> i32 {
    let mut range_timestamps = 0;
    let mut previous_block_timestamp = 0;
    let range_count = previous_blocks.len() as i32;

    for cblock in previous_blocks.values() {
        let block_data: HashMap<String, serde_json::Value> = serde_json::from_str(&to_json(cblock)).unwrap();
        if let Some(timestamp) = block_data.get("timestamp") {
            if let Some(index) = block_data.get("index") {
                if *index != 0 {
                    let nonce_time = timestamp.as_i64().unwrap() - previous_block_timestamp;
                    if nonce_time < MAX_CHANGING_INT {
                        range_timestamps += nonce_time;
                    }
                    previous_block_timestamp = timestamp.as_i64().unwrap() as i32;
                }
            }
        }
    }

    let average_block_mine = if range_count > 0 {
        range_timestamps / range_count
    } else {
        BLOCK_TIME
    };
    let minus_diff = average_block_mine / BLOCK_TIME;
    (MIN_MINING_DIFFICULTY + (previous_block_diff * MAX_CHANGING_DIFF) - (MAX_CHANGING_DIFF * minus_diff)) as i32
}

pub fn update_difficulty(current_block: &Block, previous_blocks: &HashMap<String, Block>) -> i32 {
    let mut range_timestamps = 0;
    let mut previous_block_timestamp = 0;
    let range_count = previous_blocks.len() as i32;
    let previous_block_diff = check_proof_of_work(&current_block.previous_hash);

    for cblock in previous_blocks.values() {
        let block_data: HashMap<String, serde_json::Value> = serde_json::from_str(&to_json(cblock)).unwrap();
        if let Some(timestamp) = block_data.get("timestamp") {
            if let Some(index) = block_data.get("index") {
                if *index != 0 {
                    let nonce_time = timestamp.as_i64().unwrap() - previous_block_timestamp;
                    if nonce_time < MAX_CHANGING_INT {
                        range_timestamps += nonce_time;
                    }
                    previous_block_timestamp = timestamp.as_i64().unwrap() as i32;
                }
            }
        }
    }

    if current_block.index % CHANGING_DIFF_TIME == 0 {
        let average_block_mine = if range_count > 0 {
            range_timestamps / range_count
        } else {
            BLOCK_TIME
        };
        let minus_diff = average_block_mine / BLOCK_TIME;
        return MIN_MINING_DIFFICULTY + (previous_block_diff * MAX_CHANGING_DIFF) - (MAX_CHANGING_DIFF * minus_diff);
    }
    previous_block_diff
}

pub struct Chain {
    blocks: HashMap<String, Block>,
    utxo: UTXOManager,
    head: Block,
}

impl Chain {
    pub fn new() -> Self {
        let head = Block::genesis_block();
        let mut blocks = HashMap::new();
        blocks.insert(head.hash.clone(), head.clone());
        let mut utxo = UTXOManager::new();
        for tx in &head.transactions {
            utxo.spend(tx);
        }

        Chain { blocks, utxo, head }
    }

    pub fn add_block(&mut self, next_block: Block) {
        if self.blocks.contains_key(&next_block.hash) {
            panic!("Duplicate block found when adding to chain.");
        }

        let previous_block = self.get_previous_block(&next_block);
        if previous_block.is_none() {
            panic!("New block's previous block is not in the current chain.");
        }

        let (is_verified, msg) = verify_next_block(previous_block.unwrap(), &next_block, &self.blocks);
        if !is_verified {
            panic!("New block could not be verified. Message: {}", msg);
        }

        self.blocks.insert(next_block.hash.clone(), next_block.clone());

        if next_block.index > self.head.index {
            self.update_utxo_and_head(next_block);
        }
    }

    fn update_utxo_and_head(&mut self, next_block: Block) {
        if next_block.index != self.head.index + 1 {
            panic!("Block added to block chain index is invalid.");
        }

        // Handle the fork logic (omitted for brevity)
        self.head = next_block;
    }

    fn get_previous_block(&self, current_block: &Block) -> Option<&Block> {
        self.blocks.get(&current_block.previous_hash)
        }
    }
    pub fn get_transaction(&self, transaction_hash: &str) -> Option<String> {
        for transaction_block in self.blocks.values() {
            if let Some(transaction) = transaction_block.transactions.iter().find(|tx| tx.hash == transaction_hash) {
                return Some(json!(transaction).to_string());
            }
        }
        None
    }
}

fn verify_next_block(previous_block: &block::Block, next_block: &block::Block, previous_blocks: &HashMap<String, block::Block>) -> (bool, String) {
    if next_block.index != previous_block.index + 1 {
        return (false, format!("Invalid index. Current: {}, Next: {}", previous_block.index, next_block.index));
    }

    if next_block.previous_hash != previous_block.hash {
        return (false, format!("Invalid previous hash. Current: {}, Next: {}", previous_block.hash, next_block.previous_hash));
    }

    let next_hash = block::hash_block(&previous_block.index + 1, next_block.timestamp, &next_block.transactions, next_block.noonce, &previous_block.hash);
    if next_hash != next_block.hash {
        return (false, format!("Invalid block hash. Current: {}, Expected: {}", next_block.hash, next_hash));
    }

    if next_block.index % CHANGING_DIFF_TIME != 0 {
        let previous_diff = check_proof_of_work(&next_block.previous_hash);
        let new_diff = update_difficulty(&next_block, previous_blocks);
        if new_diff != previous_diff {
            return (false, format!("Invalid proof of work. Expected diff: {}, got: {}", new_diff, previous_diff));
        }
    }

    (true, "Verified".to_string())
}

fn check_proof_of_work(block_hash: &str) -> i32 {
    block_hash.chars().count() as i32
}

File: blockchain/chain.py
import json
from typing import Dict, Tuple, List, cast, Set

import sys, os

# Add the path to the parent directory of 'blockchain'
sys.path.append(os.path.abspath('../blockchain'))
sys.path.append(os.path.abspath('../node'))
from constants import MIN_TRANSACTION_AMOUNT, COINBASE_REWARD
from constants import *
import block as block
from mine import hasProofOfWork, checkProofOfWork
import transaction as transaction
import block as chain_helper


class ChainException(Exception):
    """
    Exception class for additions to the chain
    """


class NoParentException(ChainException):
    pass


class DuplicateBlockException(ChainException):
    pass


class UTXOException(ChainException):
    pass


class UTXOManager:
    """
    The UTXO manager provides access to get referenced transactions from
    new transactions.

    It is implelmented as a dictionary mapping from transation hash to
    a tuple pair of transaction and a list of unspent indices.

    Note: None of these methods validate that the transaction's hash
    matches the corresponding data.
    """

    def __init__(self):
        self.utxo: Dict[str, Tuple[transaction.Transaction, Set[int]]] = {}

    def spend(self, newTransaction: transaction.Transaction) -> None:
        """
        Spends a transaction and updates the internal cache of utxos.

        A transaction should be verified before it is spent. If this is not
        the case, then it could create an invalid and confusing utxo state.
        Spent transactions are invalid.

        """
        for tInput in newTransaction.inputs:
            self._spendInput(tInput)

        unspentOutputIndices = set(range(len(newTransaction.outputs)))
        self.utxo[newTransaction.hash] = (newTransaction, unspentOutputIndices)

    def canSpend(
            self,
            newTransaction: transaction.Transaction) -> Tuple[bool, str]:
        """
        Verifies if a transaction can be spent based on the current UTXO cache.
        This does not account for duplicate inputs, since that can be
        done elsewhere.
        """
        inputAmounts = 0
        isCoinbase = len(newTransaction.inputs) == 0 \
                     and len(newTransaction.outputs) == 1

        for i in range(len(newTransaction.inputs)):
            tInput = newTransaction.inputs[i]
            referenced = self._getReference(tInput)
            # print(f"referenced {referenced}")
            if referenced is None:
                return False, "Referenced UTXO does not exist."

            # Verify that the signature is correct
            isValid, msg = transaction.verifyTransactionInput(
                referenced, newTransaction, i)
            if not isValid:
                return False, msg

            inputAmounts += \
                referenced.outputs[tInput.referencedOutputIndex].amount

        outputAmounts = \
            sum([output.amount for output in newTransaction.outputs])

        if not isCoinbase and inputAmounts > outputAmounts:
            return False, "Input amounts to do not match output amounts"
        elif inputAmounts == outputAmounts:
            return False, "Avoid smaller amount than output"

        return True, ""

    def revert(self, tx: transaction.Transaction):
        """
        Reverts the effect of a transaction from a UTXO.
        If the transaction has not been "spent" yet, then using this method
        may cause the internal cache to become invalid.
        """
        for tInput in tx.inputs:
            entry = self.utxo.get(tInput.referencedHash, None)
            if entry is None:
                raise UTXOException(
                    "Reference from reverted transaction does not exist.")

            unspentOutputIndices = cast(Set[int], entry[1])
            if tInput.referencedOutputIndex in unspentOutputIndices:
                raise UTXOException("Transaction index is already inspent.")

            unspentOutputIndices.add(tInput.referencedOutputIndex)

        del self.utxo[tx.hash]

    def _getReference(
            self,
            transactionInput: transaction.TransactionInput) \
            -> transaction.Transaction:
        """
        Gets a referenced transaction from the transaction input.

        When spending a transaction input - do NOT use this method.
        This does not update the internal cache of utxo objects. Instead,
        use the spend() method instead.
        """
        entry = self.utxo.get(transactionInput.referencedHash, None)
        if entry is None:
            return None

        tx = entry[0]
        # print(f"tx {tx}, entry {entry}")
        unspentOutputIndices = cast(Set[int], entry[1])

        if transactionInput.referencedOutputIndex in unspentOutputIndices:
            return tx
        else:
            print(f"entry {entry}")
            print(f"transactionInput.referenced {transactionInput.referencedOutputIndex}")
            print(f"unspentOutputIndices {unspentOutputIndices}")

        return None

    def _spendInput(self, transactionInput: transaction.TransactionInput):
        """
        Spends a UTXO. Will update the internal cache.
        """
        entry = self.utxo.get(transactionInput.referencedHash, None)
        if entry is None:
            raise UTXOException("Input can not be spent: Invalid hash.")

        tx = entry[0]
        unspentOutputIndices = cast(Set[int], entry[1])
        if transactionInput.referencedOutputIndex in unspentOutputIndices:
            unspentOutputIndices.remove(transactionInput.referencedOutputIndex)
        else:
            raise UTXOException(
                "Input can not be spent: matching " +
                "hash does not have spendable index.")


def get_update_diff(previous_block_diff: int, previous_blocks: Dict[str, block.Block]) -> int:
    range_timestamps = 0
    previous_block_timestamp = 0
    range_count = len(previous_blocks)
    
    for cblock in previous_blocks.values():
        block_data = json.loads(chain_helper.toJSON(cblock))
        timestamp = block_data.get("timestamp")
        if timestamp is not None and block_data.get("index", 0) != 0:
            nonce_time = timestamp - previous_block_timestamp
            if nonce_time < MAX_CHANGING_INT:
                range_timestamps += nonce_time
            previous_block_timestamp = timestamp
    
    average_block_mine = range_timestamps / range_count if range_count else BLOCK_TIME
    minus_diff = average_block_mine / BLOCK_TIME
    return int(MIN_MINING_DIFFICULTY + (previous_block_diff * MAX_CHANGING_DIFF) - (MAX_CHANGING_DIFF * minus_diff))



def update_difficulty(current_block: block.Block, previous_blocks: Dict[str, block.Block]) -> int:
    range_timestamps = 0
    previous_block_timestamp = 0
    range_count = len(previous_blocks)
    previous_block_diff = checkProofOfWork(current_block.previousHash)
    
    for cblock in previous_blocks.values():
        block_data = json.loads(chain_helper.toJSON(cblock))
        timestamp = block_data.get("timestamp")
        if timestamp is not None and block_data.get("index", 0) != 0:
            nonce_time = timestamp - previous_block_timestamp
            if nonce_time < MAX_CHANGING_INT:
                range_timestamps += nonce_time
            previous_block_timestamp = timestamp
    
    if current_block.index % CHANGING_DIFF_TIME == 0:
        average_block_mine = range_timestamps / range_count if range_count else BLOCK_TIME
        minus_diff = average_block_mine / BLOCK_TIME
        return int(MIN_MINING_DIFFICULTY + (previous_block_diff * MAX_CHANGING_DIFF) - (MAX_CHANGING_DIFF * minus_diff))



class Chain:
    def __init__(
            self, persistentFilename=None) -> None:
        # Blocks is a mapping from block hash to block objects
        self.blocks: Dict[str, block.Block] = {}

        # UTXO is a mapping from transaction hash to transaction objects
        self.utxo = UTXOManager()

        # The head should always point to the longest and
        # oldest chain.
        self.head = block.genesisBlock()
        self.blocks[self.head.hash] = self.head

        for tx in self.head.transactions:
            self.utxo.spend(tx)

    def addBlock(self, nextBlock: block.Block) -> None:
        """
        Adds a single block to the chain.
        """
        if nextBlock.hash in self.blocks:
            raise DuplicateBlockException(
                "Duplicate block found when adding to chain.")

        previousBlock = self.getPreviousBlock(nextBlock)
        if previousBlock is None:
            raise NoParentException(
                "New block's previous block is not in the current chain.")

        isVerified, msg = verifyNextBlock(previousBlock, nextBlock, self.blocks)
        if not isVerified:
            raise ChainException(
                "New block could not be verified." +
                "\n" + "Message: " + msg)

        # Creates a new fork in the chain if the next block's previous block
        # does exist in the current chain.
        self.blocks[nextBlock.hash] = nextBlock

        if nextBlock.index > self.head.index:
            self._updateUTXOAndHead(nextBlock)

    def _updateUTXOAndHead(self, nextBlock):
        """
        Updated the UTXO for a block being added to the end of the chain
        of the main chain.

        When a fork becomes the main chain, the UTXO from the current main chain
        are reverted and the new forks outputs are spent.
        """
        if nextBlock.index != self.head.index + 1:
            raise ChainException("Block added to block chain index is invalid")

        # Handle the fork in the rare case where the forked chain becomes
        # the main chain.
        oldChain: List[block.Block] = []
        newChain: List[block.Block] = []
        newChain.append(nextBlock)

        oldParent = self.head
        newParent = self.getPreviousBlock(nextBlock)
        while oldParent.hash != newParent.hash:
            for tx in reversed(oldParent.transactions):
                self.utxo.revert(tx)

            oldChain.append(oldParent)
            newChain.append(newParent)

            oldParent = self.getPreviousBlock(oldParent)
            newParent = self.getPreviousBlock(newParent)

        for i in range(len(newChain) - 1, -1, -1):
            transactions = newChain[i].transactions
            for j in range(len(transactions)):
                tx = transactions[j]
                canSpend, msg = self.utxo.canSpend(tx)
                if canSpend:
                    self.utxo.spend(tx)
                else:
                    # An invalid transaction was found. This means that
                    # all the new transactions need to be reverted,
                    # the old ones respent, and the invalid block deleted.

                    # Revert the transactions just added from the current block.
                    for txIndex in range(j - 1, -1, -1):
                        self.utxo.revert(transactions[txIndex])

                    # Revert the transacation from the blocks added earlier.
                    for blockIndex in range(i + 1, len(newChain)):
                        for tx in reversed(newChain[blockIndex].transactions):
                            self.utxo.revert(tx)

                    # Delete the children blocks from the invalid block
                    # as well as the :nvalid block itself from the chain.
                    for k in range(i, -1, -1):
                        del self.blocks[newChain[k].hash]

                    for oldBlock in reversed(oldChain):
                        for tx in reversed(oldBlock.transactions):
                            self.utxo.spend(tx)

                    raise UTXOException(msg)

        # If the new block increases the length of the current chain, then have
        # head point to this block. The verifyNextBlock method should check
        # that the new index is not out too large.
        self.head = nextBlock

    def addBlocks(self, newBlocks: List[block.Block]) -> None:
        """
        Adds. a list of blocks to the chain. They will be applied in
        the order of appearance in the list. In the case where the list
        is corrupt, the blocks will be removed from the chain.
        """
        for i in range(len(newBlocks)):
            newBlock = newBlocks[i]
            try:
                self.addBlock(newBlock)
            except ChainException as e:
                for j in range(i):
                    del self.blocks[newBlocks[i].hash]
                raise e

    def getChildren(self, parent: block.Block) -> List[block.Block]:
        """
        Returns all the children starting from the parent all the way
        down the chain until the head is reached. If the parent is not
        in the longest chain, then the entire longest chain is returned.

        The order of the returned children is from the oldest to the newest
        block.
        """
        if parent.index > self.head.index:
            return []

        if parent.index < 0:
            raise ChainException("Parent index is negative.")

        children: List[block.Block]
        children = []
        child = self.head
        while child.index > 0 and parent.hash != child.hash:
            children.append(child)
            child = self.getPreviousBlock(child)

        children.reverse()
        return children

    def getAncestors(self, child: block.Block, n=-1) -> List[block.Block]:
        """
        Returns an n blocks starting from the child and going
        towards the genesis block. If n is negative or unspecified,
        the entire chain will be returned. If the child block is
        None, then the head block is used instead.
        """
        longestChain: List[block.Block]
        longestChain = []
        if n == 0:
            return longestChain

        while n != 0 and child.index > 0:
            longestChain.append(child)

            child = self.getPreviousBlock(child)
            if child is None:
                raise NoParentException(
                    "Ancestors of block do not exist in chain")
            n -= 1

        return longestChain

    def getPreviousBlock(self, currentBlock: block.Block) -> block.Block:
        """
        Returns the previous block if it is in the chain
        """
        previousHash = currentBlock.previousHash
        if previousHash in self.blocks:
            return self.blocks[previousHash]

        return None

    def getTransaction(self, transaction_hash: str):
        for transaction_block in self.blocks.values():
            for data in json.loads(transaction_block.toJSON())["transactions"]:
                if data["hash"] == transaction_hash:
                    return json.dumps(data, indent=2)


def verifyNextBlock(
        previousBlock: block.Block,
        nextBlock: block.Block,
        previous_blocks: Dict[str, block.Block]) -> Tuple[bool, str]:
    """
    Verifies whether a block can syntactically can be added to the chain.
    Once a block is added to the chain with this method called, the only
    remaining check is the "canSpend" method in the UTXO.
    """

    if nextBlock.index != previousBlock.index + 1:
        return False, "Invalid index. Current: {}, Next {}".format(
            previousBlock.index, nextBlock.index)

    if nextBlock.previousHash != previousBlock.hash:
        return False, "Invalid previous hash. Current {}, Next {}".format(
            previousBlock.hash, nextBlock.previousHash)

    nextHash = block.hashBlock(
        index=previousBlock.index + 1,
        timestamp=nextBlock.timestamp,
        transactions=nextBlock.transactions,
        noonce=nextBlock.noonce,
        previousHash=previousBlock.hash)

    if nextHash != nextBlock.hash:
        return False, "Invalid block hash. Current {}, Expected {}".format(
            nextBlock.hash, nextHash)

    if nextBlock.index % CHANGING_DIFF_TIME != 0:
        previousDiff = checkProofOfWork(nextBlock.previousHash)
        if not hasProofOfWork(nextBlock.hash, previousDiff):
            return False, f"Block does not have a valid proof of work. Current diff: {previousDiff}"
    elif nextBlock.index % CHANGING_DIFF_TIME == 0:
        currentDiff = update_difficulty(nextBlock, previous_blocks)
        if not hasProofOfWork(nextBlock.hash, currentDiff):
            return False, f"Block does not have a valid proof of work. Current diff: {currentDiff}"

    # test max block count
    # print(int(MAX_SUPPLY / COINBASE_REWARD))

    if nextBlock.index > int((MAX_SUPPLY / COINBASE_REWARD)):
        return verifyMaxSupply(nextBlock.transactions)

    return verifyTransactionsSyntax(nextBlock.transactions)


def verifyMaxSupply(transactions: List[transaction.Transaction]) -> Tuple[bool, str]:
    if len(transactions) == 0 or len(transactions) > MAX_TRANSACTIONS_PER_BLOCK:
        return False, "Number of transactions is invalid."

    hasCoinbase = False

    for tx in transactions:

        # Handle coinbase
        if len(tx.inputs) == 0:
            if len(tx.outputs) == 0:
                return False, "No inputs in outputs found in transaction object."

            # Coinbase's transaction found
            if len(tx.outputs) == 1:
                if not hasCoinbase:
                    if tx.outputs[0].amount > COINBASE_REWARD_AFTER:
                        return False, "Coinbase reward is too large: {}".format(
                            tx.outputs[0].amount
                        )

                    hasCoinbase = True

                    if len(transactions) == 1:
                        return False, "Transactions only have one coinbase."
                else:
                    return False, "Multiple coinbase transactions found."
            else:
                return False, "Coinbase contains too many outputs."

    return True, ""


def verifyTransactionsSyntax(
        transactions: List[transaction.Transaction]) -> Tuple[bool, str]:
    """
    Verifies if a transaction is syntactically correct and contains
    no duplicates. This operation is pretty expensive, so it only needs
    to be run once. The list of transactions here are ones that will be
    included in a block.
    """
    if len(transactions) == 0 or len(transactions) > MAX_TRANSACTIONS_PER_BLOCK:
        return False, "Number of transactions is invalid."

    txHashes: Set[str] = set()
    referencedHashes: Set[str] = set()
    hasCoinbase = False

    for tx in transactions:
        expectedHash = transaction.Transaction.createHash(
            inputs=tx.inputs,
            outputs=tx.outputs,
            timestamp=tx.timestamp
        )

        if tx.hash != expectedHash:
            return False, \
                "Transaction hash {} does not match expected {}".format(
                    tx.hash, expectedHash)

        if tx.hash in txHashes:
            return False, "Duplicate transaction objects found. Hash: {}".format(
                tx.hash)
        else:
            txHashes.add(tx.hash)

        # Handle coinbase
        if len(tx.inputs) == 0:
            if len(tx.outputs) == 0:
                return False, "No inputs in outputs found in transaction object."

            # Coinbase's transaction found
            if len(tx.outputs) == 1:
                if not hasCoinbase:
                    if tx.outputs[0].amount > COINBASE_REWARD:
                        return False, "Coinbase reward is too large: {}".format(
                            tx.outputs[0].amount
                        )

                    hasCoinbase = True

                    if len(transactions) == 1:
                        return False, "Transactions only have one coinbase."
                else:
                    return False, "Multiple coinbase transactions found."
            else:
                return False, "Coinbase contains too many outputs."

        for tInput in tx.inputs:
            if tInput.referencedHash + str(tInput.referencedOutputIndex) in referencedHashes:
                return False, "Multiple inputs for utxo {} with index {}".format(
                    tInput.referencedHash, tInput.referencedOutputIndex)
            else:
                referencedHashes.add(tInput.referencedHash + str(tInput.referencedOutputIndex))

        for tOutput in tx.outputs:
            if tOutput.amount < MIN_TRANSACTION_AMOUNT:
                return False, \
                    "Output amount '{}' is less than the minimum reward." \
                        .format(tOutput.amount)

    return True, ""


def verifyTransactionSyntax(transactions: List[transaction.Transaction]) -> Tuple[bool, str]:
    if len(transactions) == 0 or len(transactions) > MAX_TRANSACTIONS_PER_BLOCK:
        return False, "Number of transactions is invalid."

    txHashes: Set[str] = set()
    referencedHashes: Set[str] = set()
    hasCoinbase = False

    for tx in transactions:
        expectedHash = transaction.Transaction.createHash(
            inputs=tx.inputs,
            outputs=tx.outputs,
            timestamp=tx.timestamp
        )

        if tx.hash != expectedHash:
            return False, \
                "Transaction hash {} does not match expected {}".format(
                    tx.hash, expectedHash)

        if tx.hash in txHashes:
            return False, "Duplicate transaction objects found. Hash: {}".format(
                tx.hash)
        else:
            txHashes.add(tx.hash)

        # Handle coinbase
        if len(tx.inputs) == 0:
            if len(tx.outputs) == 0:
                return False, "No inputs in outputs found in transaction object."

            # Coinbase's transaction found
            if len(tx.outputs) == 1:
                if not hasCoinbase:
                    if tx.outputs[0].amount > COINBASE_REWARD:
                        return False, "Coinbase reward is too large: {}".format(
                            tx.outputs[0].amount
                        )

                    hasCoinbase = True
                else:
                    return False, "Multiple coinbase transactions found."
            else:
                return False, "Coinbase contains too many outputs."

        for tInput in tx.inputs:
            if tInput.referencedHash + str(tInput.referencedOutputIndex) in referencedHashes:
                return False, "Multiple inputs for utxo {} with index {}".format(
                    tInput.referencedHash, tInput.referencedOutputIndex)
            else:
                referencedHashes.add(tInput.referencedHash + str(tInput.referencedOutputIndex))

        for tOutput in tx.outputs:
            if tOutput.amount < MIN_TRANSACTION_AMOUNT:
                return False, \
                    "Output amount '{}' is less than the minimum reward." \
                        .format(tOutput.amount)

    return True, ""
